Adapt CMake to import SDL2_image on windows

SandroWissmann opened the pull request:
I tried to adapt  cmake/FindSDL2_image.cmake  and CmakeLists.txt in main folder so it can properly find SDL2_image.

I also put the SDL_Image in the SDL2 image:

![Screenshot_20220920_165939](https://user-images.githubusercontent.com/42449992/191293039-a443ba28-4a1f-4980-bfbd-b694f6ce7315.png)

Now I tried to build on windows with:

```
mkdir build
cd build
cmake .. -G "MinGW Makefiles"
```

Which untortunately still givesd us this error:
![Screenshot_20220920_170132](https://user-images.githubusercontent.com/42449992/191293559-56053875-abd7-4d9c-afcf-1689d0a624e2.png)

SandroWissmann commented:
Do you maybe have an idea @crudelios ?

crudelios commented on cmake/FindSDL2_image.cmake:
```diff
@@ -36,53 +36,101 @@
+  set(SDL2_PROCESSOR_ARCH "x86")
+endif(CMAKE_SIZEOF_VOID_P EQUAL 8)
+
+GET_SDL_EXT_DIR(SDL_EXT_DIR "mixer")
```
```suggestion
GET_SDL_EXT_DIR(SDL_EXT_DIR "image")
```

Hopefully this does the trick!

SandroWissmann replied:
I changed it but this issue still gives me this:

file:///tmp/Spectacle.CRmpjV/Screenshot_20220920_180824.png

SandroWissmann replied:
unfortunatley not. See comment below ;(

SandroWissmann commented:
I tried the change but still gives this:

![Screenshot_20220920_180824](https://user-images.githubusercontent.com/42449992/191309566-b4963cc6-a153-4133-bbc5-e7e231085411.png)

crudelios commented:
That means it found the headers and not the libraries. It's an improvement!

I'll see what else needs doing.

SandroWissmann commented:
It works now. I deleted the build folder and added one by one under ext/SDL2 the files.

I then could just run 

cmake .. -G "MinGW Makefiles"

@Banderi Can you check out this branch and try to build like this:

Download:
https://github.com/libsdl-org/SDL_image/releases/download/release-2.6.2/SDL2_image-devel-2.6.2-mingw.zip

unpack it under ext/SDL2

probably you need to do the same for SDL2 and mixer.

In the end it looks like this:
![Screenshot_20220920_185939](https://user-images.githubusercontent.com/42449992/191319548-b9731c74-2bda-442d-945e-1eaf20e525bc.png)

then delete your build folder

then

mkdir build
cd build
cmake .. -G "MinGW Makefiles"

It should work now out of the box without specify LIB.

SandroWissmann commented:
actually now both ways work after delete clean the build folder:

Either with 

cmake -DCMAKE_PREFIX_PATH=LIB64 -DYSTEM_LIBS=false .. -G "MinGW Makefiles"

or just 
cmake .. -G "MinGW Makefiles"

with the fixed branch. if you supply in ext/SDL2 all SDL2 Mixer and Image.

The question which way we want to use here. Or just keep both for now and I will document it.

Banderi commented:
Works for me! Better to have it all in ext as it was originally with the other libs, so I can finally remove the LIB folder!

SandroWissmann commented:
Ok I will just add one more commit to update the readme how to build under windows / linux for other people who maybe want to build it and then from my site we could merge this. I will let you know.

SandroWissmann commented:
Ok I added one more commit with instructions how to build in the Readme. Please check it. Then you can merge it from my point of view.

Banderi commented:
I'm testing the files with the `ext` folder procedure (might need to clean up the README typos a bit btw, the links are also broken!), but it can not find SDL2_image for some reason. The rest - at least SDL2 related libs - it finds just fine:
![image](https://user-images.githubusercontent.com/16541079/191357781-9bd38fbd-896e-4b89-a69e-33932a0c7200.png)
Downloaded from [here](https://github.com/libsdl-org/SDL_image/releases/download/release-2.6.2/SDL2_image-devel-2.6.2-mingw.zip) and all together is as such:
![image](https://user-images.githubusercontent.com/16541079/191357909-5c18d878-8bf2-4514-9c7c-c88f78edb20a.png)

SandroWissmann commented:
stupid question.

Are you on the branch windows-sdl2_image-fix?


There the image cmake should be updated now and make it work.


Also delete the build folder complete and try again.

SandroWissmann commented:
My bad I forgot to push the change 

GET_SDL_EXT_DIR(SDL_EXT_DIR "image")

@crudelios suggested.

I force pushed it now. You can fetch the whole branch again.

Probably you have to reset it from origin like this:

git checkout master
git pull
git checkout windows-sdl2_image-fix
git reset --hard origin/windows-sdl2_image-fix

and then try out again. Hopefully it works. The readme I fix then tonight after work.

Banderi commented:
Cleaned up and pulled, now it finds everything and CMake generates correctly, however the build fails as it can not find `SDL_image.h`:
![image](https://user-images.githubusercontent.com/16541079/191480920-768fecb0-3d21-430a-b5fc-b365a327aebb.png)
![image](https://user-images.githubusercontent.com/16541079/191480998-56b5861f-b60c-469e-99e0-f2e8f15e2248.png)
`SDL_mixer.h` and the rest are found and included correctly where needed.

crudelios commented on cmake/FindSDL2_image.cmake:
```diff
@@ -36,53 +36,101 @@
+    file(STRINGS "${SDL2_IMAGE_INCLUDE_DIR}/SDL_image.h" SDL2_IMAGE_VERSION_MAJOR_LINE REGEX "^#define[ \t]+SDL_image_MAJOR_VERSION[ \t]+[0-9]+$")
+    file(STRINGS "${SDL2_IMAGE_INCLUDE_DIR}/SDL_image.h" SDL2_IMAGE_VERSION_MINOR_LINE REGEX "^#define[ \t]+SDL_image_MINOR_VERSION[ \t]+[0-9]+$")
+    file(STRINGS "${SDL2_IMAGE_INCLUDE_DIR}/SDL_image.h" SDL2_IMAGE_VERSION_PATCH_LINE REGEX "^#define[ \t]+SDL_image_PATCHLEVEL[ \t]+[0-9]+$")
+    string(REGEX REPLACE "^#define[ \t]+SDL_image_MAJOR_VERSION[ \t]+([0-9]+)$" "\\1" SDL2_IMAGE_VERSION_MAJOR "${SDL2_IMAGE_VERSION_MAJOR_LINE}")
```
You have a small problem on these three lines, it should be `SDL_IMAGE_MAJOR_VERSION` (the same for the `MINOR_VERSION` and `PATCHLEVEL` (notice the caps in the `image`.

crudelios reviewed (commented):
On line 755 of `CMakeLists.txt`, you need to add:

```
include_directories(${SDL2_IMAGE_INCLUDE_DIR})
```

That should fix your include issue.

Banderi commented:
Yep, now it works! -- note also that with newer SDL version it will fail to build because I haven't fixed the `USE_RENDER_GEOMETRY` functions, so just to quickly compile for the moment you need to disable it.

SandroWissmann commented:
Maybe you can just add these changes to make it work to the branch. If not I can also add them when fix the readme later today.

SandroWissmann commented:
Ok no problem for me. I will try out the branch again with your fix if it builds and after that I think you could merge it.

SandroWissmann commented on README.md:
```diff
@@ -34,3 +29,48 @@ from an unmodified game installation to run, in this case it needs _Pharaoh_ **a
+  - Important here: Choose MinGW64 
+- Install [CMake](https://github.com/Kitware/CMake/releases/download/v3.24.2/cmake-3.24.2-windows-x86_64.zip)
+- Download [SDL2](https://github.com/libsdl-org/SDL/releases/download/release-2.24.0/SDL2-devel-2.24.0-mingw.zip) and extract it to folder ext/SDL2
+- Download [SDL2_mixer](https://github.com/libsdl-org/SDL_mixer/releases/download/release-2.6.2/SDL2_mixer-devel-2.6.2-mingw.) and extract it to folder ext/SDL2
+- Download [SDL2_image](https://github.com/libsdl-org/SDL_mixer/releases/download/release-2.6.2/SDL2_mixer-devel-2.6.) and extract it to folder ext/SDL2
```
just need to remove the "." at the end of the links.

SandroWissmann commented:
perfect after this is done i will build my other pull reqeusts and also try them out on windows.
